File: app/src/main/java/com/example/padamlight/MapFragment.java
package com.example.padamlight;

import android.content.Context;
import android.graphics.Color;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.padamlight.jsonparser.DataParser;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Map Fragment
 * Responsible of displaying map and interactions with it
 */
public class MapFragment extends Fragment implements OnMapReadyCallback, MapActionsDelegate {
    private Polyline currentPolyline;

    @Nullable
    private GoogleMap mMap;

    public static MapFragment newInstance() {
        return new MapFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_map, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        // Instanciate map fragment
        SupportMapFragment mapFragment = (SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.map_fragment);
        if (mapFragment != null) {
            mapFragment.getMapAsync(this);
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        if (mMap != null) {
            mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(47.902964, 1.9092510000000402), 16f));
        }
    }

    @Override
    public void updateMap(LatLng... latLngs) {
        if (mMap != null) {
            LatLngBounds.Builder builder = new LatLngBounds.Builder();
            for (LatLng latLng : latLngs) {
                builder.include(latLng);
            }
            LatLngBounds bounds = builder.build();
            animateMapCamera(bounds);
            //FetchURL fetchURL = new FetchURL(mMap,getContext());
            //fetchURL.execute(getUrl(latLngs[0], latLngs[1], "driving"), "driving");
        }
    }

    @Override
    public void updateMarker(MarkerType type, String markerName, LatLng markerLatLng) {
        if (mMap != null) {
            MarkerOptions marker = new MarkerOptions()
                    .position(markerLatLng)
                    .title(markerName)
                    .icon(getMarkerIcon(type));

            mMap.addMarker(marker);
        }
    }

    public void onTaskDone(Object... values) {
        if (currentPolyline != null)
            currentPolyline.remove();
        currentPolyline = mMap.addPolyline((PolylineOptions) values[0]);
    }
    @Override
    public void clearMap() {
        if (mMap != null) {
            mMap.clear();
        }
    }

    private void animateMapCamera(LatLngBounds bounds) {
        if (mMap != null) {
            mMap.animateCamera(CameraUpdateFactory.newLatLngBounds(bounds, 100));
        }
    }

    private BitmapDescriptor getMarkerIcon(MarkerType type) {
        @DrawableRes int icon = R.drawable.ic_pickup;

        switch (type) {
            case PICKUP: {
                icon = R.drawable.ic_pickup;
                break;
            }
            case DROPOFF: {
                icon = R.drawable.ic_dropoff;
                break;
            }
        }
        return BitmapDescriptorFactory.fromResource(icon);
    }
    /*

    * */
    private String getUrl(LatLng origin, LatLng dest, String directionMode) {
        // Origin of route
        String str_origin = "origin=" + origin.latitude + "," + origin.longitude;
        // Destination of route
        String str_dest = "destination=" + dest.latitude + "," + dest.longitude;
        // Mode
        String mode = "mode=" + directionMode;
        // Building the parameters to the web service
        String parameters = str_origin + "&" + str_dest + "&" + mode;
        // Output format
        String output = "json";
        // Building the url to the web service
        String url = "https://maps.googleapis.com/maps/api/directions/" + output + "?" + parameters + "&key=" + getString(R.string.google_maps_key);
        return url;
    }

}
/**
 *
 *
 */

/**
 * Map interface
 * Implement this in your page where there is a map to use map methods
 */
interface MapActionsDelegate {
    void updateMap(LatLng... latLngs);

    void updateMarker(MarkerType type, String markerName, LatLng markerLatLng);

    void clearMap();
}

/**
 * Market enum
 * Define if marker is pickup type or dropoff type
 */
enum MarkerType {
    PICKUP, DROPOFF;
}

File: app/src/main/java/com/example/padamlight/views/fragments/MapFragment.kt
package com.example.padamlight.views.fragments

import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.view.ContextThemeWrapper
import androidx.fragment.app.Fragment
import com.beust.klaxon.JsonArray
import com.beust.klaxon.JsonObject
import com.beust.klaxon.Parser
import com.example.padamlight.R
import com.example.padamlight.enums.MarkerType
import com.example.padamlight.interfaces.MapActionsDelegate
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import kotlinx.android.synthetic.main.alert_dialog_itinerary.view.*
import java.net.URL
import kotlin.concurrent.thread

class MapFragment : Fragment(), OnMapReadyCallback, MapActionsDelegate {
	private var mMap: GoogleMap? = null
	private var mBuilder = LatLngBounds.Builder()

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		return inflater.inflate(R.layout.fragment_map, container, false)
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)
		// Instanciate map fragment
		val mapFragment = childFragmentManager.findFragmentById(R.id.map_fragment) as SupportMapFragment?
		mapFragment?.getMapAsync(this)
	}

	override fun onMapReady(googleMap: GoogleMap) {
		mMap = googleMap
		mMap.let {
			mMap!!.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(47.902964, 1.9092510000000402), 16f))
		}
	}

	override fun updateMap(vararg latLngs: LatLng) {
		mMap.let {
			var builder = LatLngBounds.Builder()
			for (latLng in latLngs) {
				builder.include(latLng)
			}
			mBuilder = builder
			val bounds = builder.build()
			animateMapCamera(bounds)
		}
	}

	override fun buildItinenary(from: LatLng, to: LatLng) {
		val url = getURL(from, to)
		var result: String? = ""

		val options = PolylineOptions()
		options.color(Color.RED)
		options.width(5f)

		thread {
			result = httpGet(url).toString()
		}.join()

			val parser = Parser.default()
			val stringBuilder: StringBuilder = StringBuilder(result)
			val json: JsonObject = parser.parse(stringBuilder) as JsonObject

			val routes = json.array<JsonObject>("routes")
			val points = routes!!["legs"]["steps"][0] as JsonArray<JsonObject>

			val travelTime = routes["legs"]["duration"]["text"].value.toString()

			val polypts = points.flatMap { decodePoly(it.obj("polyline")?.string("points")!!) }
			options.add(from)
			for (point in polypts) options.add(point)
			options.add(to)
			mMap!!.addPolyline(options)

			mMap!!.moveCamera(CameraUpdateFactory.newLatLngBounds(mBuilder.build(), 100))


			displayManualPickingWarning(from, to, travelTime)
	}

	private fun displayManualPickingWarning(from: LatLng, to: LatLng, travelTime: String) {
		val view = layoutInflater.inflate(R.layout.alert_dialog_itinerary, null)
		val dialog = AlertDialog.Builder(ContextThemeWrapper(view.context, R.style.AppTheme))
				.setView(view)
				.setCancelable(true)
				.create()

		view.alert_late_delivery_title.text = String.format(getString(R.string.itinerary_dialog), travelTime)
		view.alert_return.setOnClickListener {
			dialog.dismiss()
		}

		view.alert_validate.setOnClickListener {
			openInMaps(from, to)
			dialog.dismiss()
		}
		dialog.show()
	}

	private fun openInMaps(from: LatLng, to: LatLng) {
		val origin = "saddr=" + from.latitude + "," + from.longitude
		val dest = "daddr=" + to.latitude + "," + to.longitude
		val params = "$origin&$dest"

		val intent = Intent(Intent.ACTION_VIEW,
				Uri.parse("http://maps.google.com/maps?$params"))
		startActivity(intent)
	}

	private fun httpGet(myURL: String?): String? = URL(myURL).readText()

	override fun updateMarker(type: MarkerType, markerName: String, markerLatLng: LatLng) {
		if (mMap != null) {
			val marker = MarkerOptions()
					.position(markerLatLng)
					.title(markerName)
					.icon(getMarkerIcon(type))
			mMap!!.addMarker(marker)
		}
	}

	override fun clearMap() {
		if (mMap != null) {
			mMap!!.clear()
		}
	}

	private fun animateMapCamera(bounds: LatLngBounds) {
		if (mMap != null) {
			mMap!!.animateCamera(CameraUpdateFactory.newLatLngBounds(bounds, 100))
		}
	}

	private fun getURL(from: LatLng, to: LatLng): String {
		val origin = "origin=" + from.latitude + "," + from.longitude
		val dest = "destination=" + to.latitude + "," + to.longitude
		val sensor = "sensor=false"
		val key = "key=" + getString(R.string.google_maps_key)
		val params = "$origin&$dest&$sensor&$key"
		return "https://maps.googleapis.com/maps/api/directions/json?$params"
	}

	private fun getMarkerIcon(type: MarkerType): BitmapDescriptor {

		val icon = when (type) {
			MarkerType.PICKUP -> {
				R.drawable.ic_pickup
			}
			MarkerType.DROPOFF -> {
				R.drawable.ic_dropoff
			}
		}
		return BitmapDescriptorFactory.fromResource(icon)
	}

	private fun decodePoly(encoded: String): List<LatLng> {
		val poly = ArrayList<LatLng>()
		var index = 0
		val len = encoded.length
		var lat = 0
		var lng = 0

		while (index < len) {
			var b: Int
			var shift = 0
			var result = 0
			do {
				b = encoded[index++].toInt() - 63
				result = result or (b and 0x1f shl shift)
				shift += 5
			} while (b >= 0x20)
			val dlat = if (result and 1 != 0) (result shr 1).inv() else result shr 1
			lat += dlat

			shift = 0
			result = 0
			do {
				b = encoded[index++].toInt() - 63
				result = result or (b and 0x1f shl shift)
				shift += 5
			} while (b >= 0x20)
			val dlng = if (result and 1 != 0) (result shr 1).inv() else result shr 1
			lng += dlng

			val p = LatLng(lat.toDouble() / 1E5,
					lng.toDouble() / 1E5)
			poly.add(p)
		}

		return poly
	}

	companion object {
		fun newInstance(): MapFragment {
			return MapFragment()
		}
	}
}